src/vdom/component.js

```diff
@@ -20,7 +20,9 @@ export function setComponentProps(component, props, opts, context, mountAll) {
 	if ((component.__ref = props.ref)) delete props.ref;
 	if ((component.__key = props.key)) delete props.key;
 
-	if (!component.base || mountAll) {
+	if (component.constructor.getDerivedStateFromProps) {
+		component.state = extend(component.state, component.constructor.getDerivedStateFromProps(props, component.state));
+	} else if (!component.base || mountAll) {
```

Comment:
I suggest we make getDerivedStateFromProps mutually exclusive to componentWillMount and componentWillReceiveProps for two reasons:

1. It saves 2 bytes! (okay not a real reason, but just thought I'd mention it :) )
2. React makes them [mutually exclusive](https://github.com/facebook/react/blob/25dda90c1ecb0c662ab06e2c80c1ee31e0ae9d36/packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js#L654). And while Preact can differ from React, I think in this instance it might be better if follow their lead. Matching React's behavior prevents us from having to define, document, and maintain the interaction between getDerivedStateFromProps and the deprecated lifecycle methods.

Thoughts?